### PR TITLE
[dev] enable parallel PP prewarm

### PR DESCRIPTION
**Pull request by HaochenYuan**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Run a synthetic local forward/backward pass on every pipeline stage before the first scheduled training iteration. This initializes lazy kernels concurrently across PP stages without pipeline P2P dependencies.This prewarm does not capture or replay CUDA Graphs. Normal graph capture still happens later in the training schedule.   

## Performance

  On a full-scale DSV4-Flash run using 64 GB200 GPUs (`TP=1`, `PP=4`, `VPP=4`, `CP=16`, `EP=16`, 64K sequence length), startup time through the completion of the first training iteration decreased from **866.0s to 370.8s**, a **2.34x speedup**. 

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/pipeline_parallel/parallel_prewarm.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+                        micro_batch_size,
+                        rotary_pos_emb_cache,
+                    )
+                    inner_quantization_context = (
+                        nullcontext()
+                        if config.fp8 and config.fp8_recipe == Fp8Recipe.delayed
+                        else _get_quantization_context(layer, is_mtp)
+                    )
+                    with inner_quantization_context:
+                        outputs = layer.forward(*args, **kwargs)
```
Comment (yaox12) [inline megatron/core/pipeline_parallel/parallel_prewarm.py:273]: **[IMPORTANT Correctness]** `layer.forward(*args, **kwargs)` bypasses `nn.Module.__call__`, so none of the layer's forward pre/post hooks fire.

Megatron-FSDP registers its parameter-unshard forward pre-hook on the FSDP unit modules (`TransformerLayer`, `MoETransformerLayer`, `MambaLayer` by default in `mcore_fsdp_adapter.py`), and torch FSDP2 does the same. Under either wrapper this direct `.forward` runs on sharded parameters and will fail or compute garbage, and nothing in the PR rejects the FSDP configurations.

Note that `_should_call_te_cudagraph` is False at this point (no `is_graph_capturing()`, empty `cuda_graphs`), so `layer(...)` would still execute eagerly for the TE impl; only `cuda_graph_impl == "local"` would re-route through `cudagraph_manager`.

**Suggestion:** call `layer(*args, **kwargs)` so hooks run (guarding the local-cudagraph case explicitly), or assert up front that `use_megatron_fsdp` / `use_torch_fsdp2` are disabled and add a comment explaining why `.forward` is used.

File: megatron/core/transformer/transformer_config.py [hunk 2]
```diff
@@ -1106,6 +1106,18 @@ class TransformerConfig(ModelParallelConfig):
     the linear-attention kernel on the full sequence for a shard of heads. Correct but memory-heavy.
     """
 
+    #############################
+    # Pipeline Parallel Prewarm
+    #############################
+    pipeline_model_parallel_prewarm: bool = False
```
Comment (yaox12) [inline megatron/core/transformer/transformer_config.py:1112]: **[IMPORTANT Compatibility]** This new field has no CLI argument, so it is unreachable from the standard launch path.

`core_transformer_config_from_args` only copies fields for which `hasattr(args, f.name)` holds, and `arguments.py` gains no `--pipeline-model-parallel-prewarm`. Every `pretrain_gpt.py` / torchrun launch therefore leaves the flag False; only `--yaml-cfg` users can turn it on. The startup speedup reported in the PR description cannot be reproduced from this PR as-is.

**Suggestion:** add `--pipeline-model-parallel-prewarm` (`action='store_true'`) in `_add_distributed_args` so the field is populated via the normal args → `TransformerConfig` mapping.

File: megatron/core/pipeline_parallel/parallel_prewarm.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+
+        return get_fp8_context(config, layer_number)
+    if config.fp4:
+        from megatron.core.fp4_utils import get_fp4_context
+
+        return get_fp4_context(config, layer_number)
+    return nullcontext()
+
+
+def _build_packed_seq_params(config, kwargs):
```
Comment (yaox12) [inline megatron/core/pipeline_parallel/parallel_prewarm.py:71]: **[SUGGESTION]** `_build_packed_seq_params` duplicates `TransformerLayer._reconstruct_packed_seq_params_from_kwargs`, which `TransformerLayer.forward` already calls whenever `cu_seqlens_q` is present in kwargs (same `max_seqlen` derivation, same `pad_between_seqs=True`).

Two copies of this reconstruction will drift over time. Mamba layers never receive `cu_seqlens_*` from the base `get_layer_static_inputs`, so they would be unaffected by removing this helper.

**Suggestion:** drop `_build_packed_seq_params` and pass the `cu_seqlens_*` kwargs straight through to the layer.

File: megatron/core/pipeline_parallel/parallel_prewarm.py [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        return tensors
+    if isinstance(value, (tuple, list)):
+        tensors = []
+        for item in value:
+            tensors.extend(_collect_differentiable_tensors(item))
+        return tensors
+    return []
+
+
+def _discover_local_layers(model):
```
Comment (yaox12) [inline megatron/core/pipeline_parallel/parallel_prewarm.py:37]: **[SUGGESTION]** `_discover_local_layers` silently skips any chunk without a `decoder` attribute, and THD-shaped inputs are only produced when `sequence_packing_scheduler` is set.

Multimodal models (`language_model.decoder`) end up prewarming zero layers, and plain packed-THD datasets prewarm SBHD attention kernels while the real run uses THD kernels. The only signal today is an INFO log saying "0 local layers".

**Suggestion:** emit a warning when no local layers are discovered, and document the THD gating in the `pipeline_model_parallel_prewarm` docstring.

File: megatron/core/pipeline_parallel/parallel_prewarm.py [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    reset_model_temporary_tensors(config, model)
+
+
+def prewarm_pipeline_model_parallel(
+    model,
+    config,
+    seq_length,
+    micro_batch_size,
+    optimizers=(),
+    pg_collection=None,
```
Comment (yaox12) [inline megatron/core/pipeline_parallel/parallel_prewarm.py:148]: **[SUGGESTION]** No tests cover this function, and the `pipeline_model_parallel_size > 1` assert in `TransformerConfig.__post_init__` makes single-GPU coverage impossible.

The backup/restore of FP8 scaling tensors, module buffers, `is_first_microbatch`, RNG state, loss-scaler class attributes, and grad buffers is exactly the kind of state that regresses silently. A test asserting that model buffers, RNG state, and `main_grad` are bitwise identical before and after the call would guard it.

**Suggestion:** add a unit test (2-GPU PP=2, or fake-pg) and relax the PP>1 assert to a warning so the function can be exercised with PP=1.

File: megatron/core/transformer/transformer_layer.py [hunk 6]
```diff
@@ -1276,7 +1286,7 @@ def get_layer_static_inputs(self, seq_length, micro_batch_size):
             static_inputs["padding_mask"] = torch.zeros(
                 1, slen_for_mask, dtype=torch.bool, device=device
             )
-        elif attn_in_graph:
+        elif attention_is_executed:
```
Comment (yaox12) [inline megatron/core/transformer/transformer_layer.py:1289]: **[IMPORTANT Performance]** With `for_pipeline_prewarm=True`, `attention_is_executed` is always True, so the SBHD branch below always builds the synthetic `attention_mask`: a `(seq_length/cp) x seq_length` fp32 tensor created on CPU, moved to GPU, then tiled by micro-batch size.

`create_attention_mask_in_dataloader` defaults to True, so this is the common case. At the 64K sequence lengths the prewarm targets that is ~16 GB of host fp32 plus ~4 GB of device bool per microbatch, before the first iteration runs. The CUDA-graph capture path had the same code, but the prewarm now reaches it for every long-sequence PP run that is not using a packing scheduler.

**Suggestion:** build the mask directly on device as bool, e.g.
```python
~torch.tril(torch.ones((slen_per_cp, seq_length), dtype=torch.bool, device=device))
```
and skip it entirely when `attn_mask_type` is causal and the attention backend ignores the explicit mask.
